File: Data Structures & Algorithms/delete-node-in-a-bst/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} key
     * @return {TreeNode}
     */
    findMin(node) {
        if(!node){
            return null
        }
        if(node.left){
            return this.findMin(node.left)
        }else{
            return node.val
        }

    }
    deleteNode(root, key) {
        if(!root){
            return null
        }
        if(root.val == key){
            let min = null;
            if(root.right == null && root.left == null){
                return null
            }
            if(root.right != null){
               min = this.findMin(root.right)
               root.val = min
               root.right = this.deleteNode(root.right, min)
            }else{
                min = this.findMin(root.left)
                root.val = min
                root.left = this.deleteNode(root.left, min)
            }
        }else{
            if(root.val > key){
                root.left = this.deleteNode(root.left, key)
            }else{
                root.right = this.deleteNode(root.right, key)
            }
            
        }

        return root
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        res = True
        def dfs(node, lower_limit, upper_limit):
            nonlocal res
            if not res or not node:
                return 
            
            if node.val >= upper_limit or node.val <= lower_limit:
                res = False
                return 
            
            dfs(node.left, lower_limit, node.val)
            dfs(node.right, node.val, upper_limit)
            return
        
        dfs(root, float("-inf"), float("inf"))
        return res